loader/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/loader/BadRow.scala

```diff
@@ -12,52 +12,150 @@
  */
 package com.snowplowanalytics.snowplow.storage.bigquery.loader
 
-import java.util.Base64
-
-import cats.implicits._
 import cats.data.NonEmptyList
 
-import io.circe.{Decoder, DecodingFailure, Encoder, Json}
+import io.circe.{Encoder, Json}
 import io.circe.syntax._
 
+import com.snowplowanalytics.iglu.client.ClientError
+import com.snowplowanalytics.iglu.schemaddl.bigquery.{CastError => BigQueryCastError}
+import com.snowplowanalytics.iglu.core.{SchemaKey, SelfDescribingData}
+import com.snowplowanalytics.iglu.core.circe.instances._
+
+import com.snowplowanalytics.snowplow.analytics.scalasdk.Event
+
+import com.snowplowanalytics.snowplow.storage.bigquery.common.{BadRowSchemas, ProcessorInfo}
+
+
 /**
-  * Class representing an enriched event that failed BigQuery Loader transformation,
-  * e.g. due unavailable schema, bug in Schema DDL or unexpected structure
-  * Unlike "failed insert" it did not get to BigQuery table
+  * Represents events which are problematic to process due to
+  * several reasons such as could not be parsed, error while validate
+  * them with Iglu, error while casting them to BigQuery types etc.
   */
-case class BadRow(line: String, errors: NonEmptyList[String]) {
-  def compact: String = this.asJson.noSpaces
+sealed trait BadRow {
+  def getSelfDescribingData: SelfDescribingData[Json]
+  def compact: String  = getSelfDescribingData.asJson.noSpaces
 }
 
 object BadRow {
-  private val Line = "line"
-  private val Errors = "errors"
-
-  implicit val encoder: Encoder[BadRow] =
-    Encoder.instance { badRow: BadRow =>
-      val encoded = Base64.getEncoder.encode(badRow.line.getBytes)
-      Json.fromFields(List(
-        (Line, Json.fromString(new String(encoded))),
-        (Errors, Json.fromValues(badRow.errors.toList.map(Json.fromString)))
-      ))
-    }
+  /**
+    * Represents the failure case where data can not be parsed as a proper event
+    *
+    * @param payload data blob tried to be parsed
+    * @param errors  errors in the end of the parsing process
+    */
+  final case class ParsingError(payload: String, errors: NonEmptyList[String]) extends BadRow {
+    def getSelfDescribingData: SelfDescribingData[Json] =
+      SelfDescribingData(BadRowSchemas.LoaderParsingError, (this: BadRow).asJson)
+  }
+
+  /**
+    * Represents the errors due to something went wrong internally
+    * In this kind of errors, event is parsed properly, therefore
+    * enriched event is given in the error as properly parsed
+    *
+    * @param enrichedEvent event which is enriched successfully
+    * @param errors        info of errors
```

Comment:
Nit: this kind of ScalaDoc formatting (with gap between param and description) is slightly unusual for our projects.